### PR TITLE
Modify data-security to respect OpenEdu methodology

### PR DESCRIPTION
**Pull request by Jetoson**
# Prerequisite Checklist

- [x] Read the [contribution guidelines](https://github.com/open-education-hub/ccas-internal/blob/main/CONTRIBUTING.md#pull-requests) regarding submitting new changes to the project;
- [x] Tested your changes against relevant architectures and platforms;
- [x] Updated relevant documentation (if needed).

### CODE REVIEW

File: data-security/slides/Makefile [hunk 1]
```diff
@@ -0,0 +1,2 @@
+include ../../../../common/slides/Makefile
+
```
Comment (teodutu) [inline data-security/slides/Makefile]:
```suggestion
```

File: data-security/reading/README.md [hunk 2]
```diff
@@ -0,0 +1,246 @@
+---
+linkTitle: 07. Data Security
+type: docs
+weight: 10
+---
+
```
Comment (teodutu) [inline data-security/reading/README.md]:
```suggestion
```
This was only needed for our older rendering engine, but not for Docusaurus.

File: data-security/reading/README.md [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+# Data Security
+
+## Table of Contents
+
+Use [gh-md-toc](https://github.com/ekalinin/github-markdown-toc).
+
```
Comment (teodutu) [inline data-security/reading/README.md]:
```suggestion
```

File: data-security/reading/README.md [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+Without complex mathematics and algorithms people found intuitive ways to codify their messages.
+
+Classical ciphers were of two types: substitution and transposition of letters.
+
+## Caesar Cipher
+
+One of the most known and simple classical ciphers, named after the great Roman conqueror Julius Caesar, who also used it to communicate with his crush Cleopatra, works by using the alphabet substition of a letter by a given shift.
+
+For example if we set our shift to +5 we can see how the cipher works in the following drawing ![Caesar Cipher Example](../media/caesar-cipher.svg)
+
```
Comment (teodutu) [inline data-security/reading/README.md]: Move this content to `introduction.md`.

File: data-security/reading/README.md [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+But how exactly does OTP work?
+
+Given a plaintext that can be translated into binary data and the previously mentioned pad made up of random and unpredictable bits, the attacker won't have any clue about the plaintext when they see the ciphertext.
+
+![XOR Attack](../media/xor-attack.svg)
+
+You will probably think "Why don't we use OTP's if it's impossible for an attacker to decipher it if the OTP's rules are implemented correctly?".
+The problem is OTP is impractical: if the data you want to transmit is big then the pad must be at least the size of the data.
+Also you'll have to transmit the pad to other parties ahead of time in order to get it deciphered in practical time.
+
```
Comment (teodutu) [inline data-security/reading/README.md]: Move this content to `xor-encrpytion.md`.

File: data-security/reading/README.md [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+print(f"Key: {KEY}")
+print(f"IV: {cipher.iv}") # The library will create a random IV if not specified when we create the AES object 
+print(f"Ciphertext: {ciphertext}")
+
+cipher = AES.new(KEY, AES.MODE_CBC, cipher.iv) # we initialize a new object specifying the IV used in the encryption
+
+decrypted_plaintext = cipher.decrypt(ciphertext) 
+print(f"Plaintext: {decrypted_plaintext}") 
+```
+
```
Comment (teodutu) [inline data-security/reading/README.md]: Move this content to `symmetric-cryptography.md`.

File: data-security/reading/README.md [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+It heavily relies on the computational difficulty of factoring large numbers to ensure secure communication.  
+RSA uses a pair of keys, a public key for encryption and a private key for decryption.  
+We will work with the form of RSA with keys represented by numbers.  
+RSA is extensively employed for secure online transactions, digital signatures, secure email communication, and authentication protocols.  
+Its strength lies in its resistance to brute-force attacks, making it a fundamental component of secure communication in today's interconnected world.
+
+## Activities
+
+Today's activites are a bit different than usual. They will contain encrypted strings and the required keys for decryption. The proccess is about trying to understand how the decryption works when you have everything needed and only then you shall seek to perform attacks on various encrypted texts.
+
```
Comment (teodutu) [inline data-security/reading/README.md]:
```suggestion
```

File: data-security/reading/README.md [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+## Further Reading
```
Comment (teodutu) [inline data-security/reading/README.md]: Either add some interesting links or remove this section. Don't leave it as a stub.

File: config.yaml [hunk 9]
```diff
@@ -94,8 +94,20 @@ docusaurus:
...
                 - Reading: reading/README.md
+                - Drills:
+                    path: drills/
+                    subsections:
+                      - Tasks:
+                          path: tasks/
+                          subsections:
+                            - AES CBC/: aes-cbc/README.md
+                            - Caesar Cipher/: caesar-cipher/README.md
+                            - Xor Encryption/: xor-encryption/README.md
```
Comment (teodutu) [inline config.yaml]:
```suggestion
                            - Caesar Cipher/: caesar-cipher/README.md
                            - Xor Encryption/: xor-encryption/README.md
                            - AES CBC/: aes-cbc/README.md
```

**Review by teodutu — CHANGES_REQUESTED** [review]
I made an initial pass and given some general feedback. Apart from it, fix the failing CI/CD tests (including deployment) [1]. To fix spellcheck, you'll have to add some words or acronyms to these lists [2]. Make another PR to that repo with the added words. Make sure to run `sort -u -o <file_name> <file_name>`.

[1] https://github.com/security-summer-school/essentials-security/pull/4/checks
[2] https://github.com/open-education-hub/actions/tree/main/spellcheck/config

**Comment by Jetoson** [comment]
I made the requested changes and opened a PR[1] to fix the failing spellcheck test.

1. https://github.com/open-education-hub/actions/pull/59

**Review by teodutu — CHANGES_REQUESTED** [review]
Watch out for the super-linter [1] and checkpatch [2] errors. To fix your commit description:

1. First format its title as described in the commit guidelines [3]: `chapters/data-security: Add session content`.
2. Add a description about what the content is about. Use the example commit at [3] as an example.
3. Sign your commit using `git commit --amend -s`.

[1] https://github.com/security-summer-school/essentials-security/actions/runs/9562547309/job/26372997637?pr=4#step:3:989
[2] https://github.com/security-summer-school/essentials-security/actions/runs/9562547309/job/26372998333?pr=4#step:3:534
[3] https://github.com/security-summer-school/essentials-security/blob/main/CONTRIBUTING.md#commits

File: data-security/reading/symmetric-cryptography.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+print(f"Key: {KEY}")
+print(f"IV: {cipher.iv}") # The library will create a random IV if not specified when we create the AES
+object 
+print(f"Ciphertext: {ciphertext}")
+
+cipher = AES.new(KEY, AES.MODE_CBC, cipher.iv) # we initialize a new object specifying the IV used in the encryption
+
+decrypted_plaintext = cipher.decrypt(ciphertext)
+print(f"Plaintext: {decrypted_plaintext}")
+``
```
Comment (teodutu) [inline data-security/reading/symmetric-cryptography.md]:
```suggestion
```
```

File: data-security/reading/symmetric-cryptography.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+from Crypto.Util.Padding import pad
+
+KEY = get_random_bytes(16) # generate a random 16 bytes key
+
+plaintext = b"$ecret_1nformat1on"
+
+cipher = AES.new(KEY, AES.MODE_CBC) #
+ciphertext = cipher.encrypt(pad(plaintext, AES.block_size))
+
+#Encryption
```
Comment (teodutu) [inline data-security/reading/symmetric-cryptography.md]:
```suggestion
# Encryption
```

**Review by teodutu — CHANGES_REQUESTED** [review]
PR rendering fails because you placed the content directly in the repo's root. Instead, the `data-securty` chapter should be placed in `chapters/data/`. Then you need to remove all `overview` sections from `config.yaml`.

After that, take a look at the automatic tests that are failing [1] [2] and fix them. To fix the checkpatch errors, you'll first have to squash your commits like so:
1. run `git rebase -i main`. It will open your default editor and list all the commits on your branch.
2. Change the first word of all lines except for the first one from `pick` to `s` (from "squash") and save and exit.
3. A new editor tab will open with the commit messages. Reword them like so:
```
data-security: Add session content

Then add a short description of the content
```
4. Run `git commit --amend -s --no-edit` to add your commit signature
5. Finally, run `git push -f` to push your changes upstream.

Amend this new commit for every new revision using `git commit --amend --no-edit`.

[1] https://github.com/security-summer-school/essentials-security/actions/runs/9572698155/job/26392592312?pr=4
[2] https://github.com/security-summer-school/essentials-security/actions/runs/9572698155/job/26392591945?pr=4#step:3:995
[3] https://github.com/security-summer-school/essentials-security/actions/runs/9572698155/job/26392592540?pr=4

**Comment by Jetoson** [comment]
Made the requested changes and fixed all `checkpatch` and `super-linter` errors. Though the spellcheck tests has passed, the PR[1] with few additional words is still open.

1. https://github.com/open-education-hub/actions/pull/60/

File: chapters/data/data-security/drills/tasks/aes-cbc/README.md [hunk 12]
```diff
@@ -0,0 +1,8 @@
+# AES CBC
+
```
Comment (teodutu) [inline chapters/data/data-security/drills/tasks/aes-cbc/README.md:2]:
```suggestion

Decrypt the string below using the provided key and IV.

```
Describe the task a little bit so it's not so abrupt.

File: chapters/data/data-security/drills/tasks/caesar-cipher/README.md [hunk 13]
```diff
@@ -0,0 +1,6 @@
+# Caesar Cipher
+
+Encrypted string: `JJJ{nyp_1j_dp_j4cru_jy1w7vu}`
```
Comment (teodutu) [inline chapters/data/data-security/drills/tasks/caesar-cipher/README.md]:
```suggestion
Decrypt the following string: `JJJ{nyp_1j_dp_j4cru_jy1w7vu}`
```

File: chapters/data/data-security/drills/tasks/xor-encryption/README.md [hunk 14]
```diff
@@ -0,0 +1,6 @@
+# Xor Encryption
+
+Encrypted string: `0x3b3c241f180d4715275b2d332a2b181a015f185a34111f5e041012`
```
Comment (teodutu) [inline chapters/data/data-security/drills/tasks/xor-encryption/README.md]:
```suggestion
Decrypt the following string: `0x3b3c241f180d4715275b2d332a2b181a015f185a34111f5e041012`
```

File: chapters/data/data-security/reading/symmetric-cryptography.md [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+
+![AES](../media/aes.svg)
+
+## Block cipher mode of operation
+
+Block ciphers seems to solve the OTP key size problem but "what if we want to encrypt more than a block of
+data?"
+We will need to use a block cipher mode of operation.
+The mode of operation describes how to securely apply the cipher's single-block operation to transform data
+larger than the block size.
```
Comment (teodutu) [inline chapters/data/data-security/reading/symmetric-cryptography.md]:
```suggestion
Block ciphers seems to solve the OTP key size problem but "what if we want to encrypt more than a block of data?"
We will need to use a block cipher mode of operation.
The mode of operation describes how to securely apply the cipher's single-block operation to transform data larger than the block size.
```

File: chapters/data/data-security/reading/symmetric-cryptography.md [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+### ECB
+
+ECB stands for Electronic Codebook and is the simplest and not to be used encryption mode.
+The way it works it's straightforward: the message is divided into multiple blocks and each block is
+encrypted separately.
```
Comment (teodutu) [inline chapters/data/data-security/reading/symmetric-cryptography.md]:
```suggestion
The way it works it's straightforward: the message is divided into multiple blocks and each block is encrypted separately.
```

File: chapters/data/data-security/reading/symmetric-cryptography.md [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+
+![ECB](../media/ecb.svg)
+
+The main security issue with ECB mode is that encrypting the same block of plaintext always returns the same
+block of ciphertext.
```
Comment (teodutu) [inline chapters/data/data-security/reading/symmetric-cryptography.md]:
```suggestion
The main security issue with ECB mode is that encrypting the same block of plaintext always returns the same block of ciphertext.
```

File: chapters/data/data-security/reading/xor-encrpytion.md [hunk 18]
```diff
@@ -0,0 +1,86 @@
+# Exclusive Or (XOR)
+
+Exclusive or, also called "XOR" is a binary operator that behave like an programmable inverter so as the
+output is true if either first or second input is true but not both.
```
Comment (teodutu) [inline chapters/data/data-security/reading/xor-encrpytion.md]:
```suggestion
Exclusive or, also called "XOR" is a binary operator that behave like an programmable inverter so as the output is true if either first or second input is true but not both.
```

File: chapters/data/data-security/reading/xor-encrpytion.md [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+def xor_strings(w1, w2):
+        return "".join([chr(ord(a) ^ ord(b)) for a,b in zip(w1, w2)])
+
+print(xor_strings("111","bbb").encode())
+```
+
+## OTP (One Time Pad)
+
+At a first glance, XOR may look too simple to provide a secure encryption scheme but it is the basis of the
+most secure encryption algorithm, called a **one-time pad**.
```
Comment (teodutu) [inline chapters/data/data-security/reading/xor-encrpytion.md]:
```suggestion
At a first glance, XOR may look too simple to provide a secure encryption scheme but it is the basis of the most secure encryption algorithm, called a **one-time pad**.
```

File: chapters/data/data-security/reading/xor-encrpytion.md [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+The name suggests using a random sequence of bits called `pad` with which we XOR each bit in the plaintext.
+The security is guaranteed by the fact that the pad must be only used once therefore the name.
+
+But how exactly does OTP work?
+
+Given a plaintext that can be translated into binary data and the previously mentioned pad made up of random
+and unpredictable bits, the attacker won't have any clue about the plaintext when they see the ciphertext.
```
Comment (teodutu) [inline chapters/data/data-security/reading/xor-encrpytion.md]:
```suggestion
Given a plaintext that can be translated into binary data and the previously mentioned pad made up of random and unpredictable bits, the attacker won't have any clue about the plaintext when they see the ciphertext.
```

File: chapters/data/data-security/reading/xor-encrpytion.md [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+
+![XOR Attack](../media/xor-attack.svg)
+
+You will probably think "Why don't we use OTP's if it's impossible for an attacker to decipher it if the
+OTP's rules are implemented correctly?"
+The problem is OTP is impractical: if the data you want to transmit is big then the pad must be at least the
+size of the data.
+Also you'll have to transmit the pad to other parties ahead of time in order to get it deciphered in
+practical time.
```
Comment (teodutu) [inline chapters/data/data-security/reading/xor-encrpytion.md]:
```suggestion
You will probably think "Why don't we use OTP's if it's impossible for an attacker to decipher it if the OTP's rules are implemented correctly?"
The problem is OTP is impractical: if the data you want to transmit is big then the pad must be at least the size of the data.
Also you'll have to transmit the pad to other parties ahead of time in order to get it deciphered in practical time.
```

File: chapters/data/data-security/reading/xor-encrpytion.md [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+|---|---|
+|  a ⊕ (b ⊕ c) = (a ⊕ b) ⊕ c  | XOR can be applied in any order|
+|  a ⊕ a = 0 | XORing a bit with itself is always 0|
+|  a ⊕ 0 = a | XORing a bit with 0 doesn't change the bit value|
+|  a ⊕ b = b ⊕ a  |You can flip the operands order|
+
+## XOR Encryption
+
+We can combine the previously mentioned properties in order to obtain a cool property that's used for
+decryption:
```
Comment (teodutu) [inline chapters/data/data-security/reading/xor-encrpytion.md]:
```suggestion
We can combine the previously mentioned properties in order to obtain a cool property that's used for decryption:
```

File: chapters/data/data-security/reading/xor-encrpytion.md [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+$a ⊕ b ⊕ a = b$
+
+The first XOR between $a$ and $b$ is the encryption part whereas the second XOR operations marks the
+decryption phase:
```
Comment (teodutu) [inline chapters/data/data-security/reading/xor-encrpytion.md]:
```suggestion
The first XOR between $a$ and $b$ is the encryption part whereas the second XOR operations marks the decryption phase:
```

File: chapters/data/data-security/reading/xor-encrpytion.md [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+$
+ciphertext = plaintext ⊕ key\\
+ciphertext ⊕ key  = plaintext ⊕ key ⊕ key = plaintext
+$
+
+We only defined the XOR as working on single bits, so in order to be able to use the XOR for encryption we
+need to implement a bitwise XOR function that can operate on more than one bit.
+
+Luckily for us, Python already provide us with the `^` (caret) operator that can be applied to integers to
+XOR them.
```
Comment (teodutu) [inline chapters/data/data-security/reading/xor-encrpytion.md]:
```suggestion
We only defined the XOR as working on single bits, so to be able to use the XOR for encryption we need to implement a bitwise XOR function that can operate on more than one bit.

Luckily, Python already provides us with the `^` (caret) operator that can be applied to integers to XOR them.
```

**Review by teodutu — CHANGES_REQUESTED** [review]
The content is good now. I made some small suggestions about sentence formatting. In addition, building the website fails because of the slides. Remove all references to `data-security/slides` from `config.yaml`.

**Review by teodutu — APPROVED** [review]
:+1: GG!

**Comment by teodutu** [comment]
@RaresCroicia add your review so we can merge this after it's addressed.

**Review by RaresCroicia — APPROVED** [review]
Seems good to me!

**Pull request merged**